src/quarto/ai/minimax.go

```diff
@@ -8,66 +8,143 @@ import (
 	"time"
 )
 
+const UndefinedValue = -1
+const WinningLeafValue = 0
+const LoosingLeafValue = 100
+
 // StateNode define data for a node a the tree used in minmax algorithm
 type StateNode struct {
 	State  state.State
 	Value  int
+	MyNode bool
 	Childs []StateNode
 }
 
 // InitAllTree creates tree of possibilities
-func InitAllTree(currentState state.State, quit chan struct{}) StateNode {
+func InitAllTree(currentState state.State, piecesList []int, boxList []grid.Point, quit chan struct{}) StateNode {
 	nextState := state.CopyState(currentState)
-	tree := InitNode(nextState)
-	tree = AppendChildNodes(tree, 2, quit)
+	tree := InitNode(nextState, false)
+
+	tree = AppendChildNodes(tree, 15, piecesList, boxList, quit)
 	return tree
 }
 
 // InitNode creates new node
-func InitNode(currentState state.State) StateNode {
-	return StateNode{State: currentState, Value: 0}
+func InitNode(currentState state.State, myNode bool) StateNode {
+	return StateNode{State: currentState, Value: UndefinedValue, MyNode: myNode}
 }
 
-// AppendChildNodes creates nodes of current node
-func AppendChildNodes(node StateNode, depth int, quit chan struct{}) StateNode {
+// AppendChildNodes creates child nodes of current node
+func AppendChildNodes(node StateNode, depth int, piecesList []int, boxList []grid.Point, quit chan struct{}) StateNode {
 	select {
 		case <-quit:
 			return node
 		default:
 			if depth >= 0 {
 				node.Childs = []StateNode{}
-				piecesList := state.GetRemainingPiecesListFromState(node.State)
-				boxList := grid.GetEmptyBoxes(node.State.Grid)
-				if len(piecesList) > 0 && len(boxList) > 0 {
-					for j := 0; j < len(piecesList); j++ {
-						if node.State.Piece == 0 {
-							nextState := state.CopyState(node.State)
-							nextState.Piece = piecesList[j]
-							childNode := InitNode(nextState)
-							node.Childs = append(node.Childs, AppendChildNodes(childNode, depth-1, quit))
-						} else {
-							for i := 0; i < len(boxList); i++ {
+				if len(piecesList) > 1 && len(boxList) > 1 {
```

Comment:
You should split the code into small functions because it's not so readable. And in addition you could test small part of code.